### PR TITLE
Remove legacy per-request JSON spill shard merge path

### PR DESCRIPTION
**Pull request by alexwitherspoon**
## Summary

- Drops read-side support for pre-JSONL `{pid}_{hex}.json` spill shards now that #116 writes only per-worker JSONL journals.
- Removes deprecated `getMetricsSpillSnapshotPath()` and legacy glob/parse merge branches.
- Simplifies aggregator to journal-only merge and counts `spills_deleted` when a journal is successfully consumed.
- Keeps **orphan prune** for stale `.json` shards (and `.jsonl` / `.merging` artifacts) as a deploy safety net - they are not merged, only deleted when older than `METRICS_SPILL_ORPHAN_MAX_AGE_SECONDS`.
- Updates tests, operator docs, and scheduler/CLI comments to journal terminology.

## Depends on

- #116 (`feat/metrics-spill-jsonl`) must merge first. This PR targets that branch so the diff is cleanup-only.

## Deploy note

Merge only after #116 has been in production long enough that `cache/metrics/spill/` has no remaining `.json` shards (or you accept losing unmigrated stragglers). Check:

```bash
find /var/www/html/cache/metrics/spill -name '*.json' | wc -l
```

## Test plan

- [x] `make test-ci`
- [x] After #116 merge: rebase this branch onto `main` (or change PR base) before final merge
- [x] Post-deploy: confirm spill dirs contain only `{pid}.jsonl` and hourly metrics continue updating

### CODE REVIEW

File: lib/metrics-spill-aggregator.php [hunk 1]
```diff
@@ -343,8 +287,7 @@ function metrics_spill_aggregator_prune_orphan_spills(array &$stats, $t0Ns): voi
             if (strpos($name, '.tmp.') !== false) {
                 continue;
             }
-            $isSpillArtifact = str_ends_with($name, '.json')
-                || str_ends_with($name, '.jsonl')
+            $isSpillArtifact = str_ends_with($name, '.jsonl')
                 || strpos($name, '.merging.') !== false;
             if (!$isSpillArtifact) {
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-aggregator.php]: Agreed on the safety net. Restored .json in the orphan-prune filter without re-adding merge support so pre-JSONL stragglers age out instead of living forever. Fixed in e9f14b0.

File: lib/metrics-spill-aggregator.php [hunk 2]
```diff
@@ -117,22 +116,15 @@ function metrics_run_spill_aggregator_once(): array
...
                 }
 
-                if (!$isJournal) {
-                    $pendingDeletes[] = $spillPath;
-                }
                 $hourHadMerges = true;
                 $filesMergedThisRun++;
                 $stats['spills_merged'] += $mergedUnits;
+                $stats['spills_deleted']++;
             }
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-aggregator.php]: Good catch - claim success with zero valid lines was indistinguishable from claim failure and did not advance the file budget or spills_deleted. merge_journal now returns 0 vs null, empty claims are deleted immediately, and empty hour dirs are rmdir'd. Fixed in c9d2b2a (includes the @return contract update).

File: lib/metrics-spill-aggregator.php [hunk 3]
```diff
@@ -226,89 +212,47 @@ function metrics_spill_aggregator_write_hour_bucket(
...
+ * @param string               $journalPath Absolute path to {pid}.jsonl
+ * @param string               $hourId      UTC hour bucket id
+ * @param array<string, mixed> $hourData    Hourly bucket (mutated in place)
+ * @return int|null Number of journal lines merged, or null when claim/merge produced nothing
  */
-function metrics_spill_aggregator_merge_spill_path(
-    string $spillPath,
+function metrics_spill_aggregator_merge_journal(
+    string $journalPath,
     string $hourId,
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-aggregator.php]: Agreed - that @return conflation was part of the budget/telemetry gap above. merge_journal now returns null only on claim/path failure and 0 when claimed with no valid lines. Docblock updated in c9d2b2a.

File: lib/metrics-spill-payload.php [hunk 4]
```diff
@@ -11,9 +11,9 @@
 /**
  * Validate decoded spill JSON and return normalized counter keys for hourly merge.
  *
- * Used by the spill aggregator after reading a shard file; rejects schema mismatch, hour mismatch,
+ * Used by the spill aggregator after decoding a journal line; rejects schema mismatch, hour mismatch,
  * non-numeric counter values, empty counter maps, unreadable keys, or any counter name not recognized
- * by {@see metrics_flat_counter_key_is_recognized()} so corrupt or forward-incompatible shards are not merged
+ * by {@see metrics_flat_counter_key_is_recognized()} so corrupt or forward-incompatible lines are not merged
  * and are left on disk for inspection.
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-payload.php]: You're right - that line was stale after the journal merge path. Updated the docblock to say invalid lines are skipped rather than left for inspection. Fixed in 2nd commit on this thread (payload doc commit: 556470e).

[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-payload.php]: Fair point on the mixed-line case - 556470e fixed the stale "left on disk" wording, and 5ab55cc adds that invalid lines are not preserved when a journal with valid lines is consumed and deleted after merge.

File: tests/Unit/MetricsSpillAggregatorTest.php [hunk 5]
```diff
@@ -9,46 +9,236 @@
...
+        foreach (glob($hourDir . '/*') ?: [] as $file) {
+            @unlink($file);
+        }
+        @rmdir($hourDir);
+
+        @unlink(getMetricsAggregatorLastRunPath());
+        @unlink(getMetricsAggregatorLockPath());
+    }
+
+    public function testMerge_SingleJournalLineIntoHourlyFile(): void
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillAggregatorTest.php]: Agreed. Renamed the aggregator integration tests to testFunctionName_Scenario_ExpectedBehavior per CODE_STYLE.md in 0a84dcb.

File: scripts/unified-webcam-worker.php [hunk 6]
```diff
@@ -266,7 +266,7 @@ function runWorkerMode(string $airportId, int $camIndex): int
         $worker = WebcamWorkerFactory::create($airportId, $camIndex);
         $result = $worker->run();
 
-        // Best-effort: workers do not use the FPM shutdown spill hook; write a CLI spill shard for the scheduler merge.
+        // Best-effort: workers do not use the FPM shutdown spill hook; append a CLI spill journal line for the scheduler merge.
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply scripts/unified-webcam-worker.php:269]: Good call. Renamed the helper to metrics_write_spill_journal_and_reset_counters() in 4e0b73d so the call site matches what it does now.

File: scripts/unified-webcam-worker.php [hunk 7]
```diff
@@ -329,7 +329,7 @@ function runSingleMode(string $airportId, int $camIndex): int
         echo "  Duration: {$elapsed}ms\n";
         echo "\n";
 
-        // Flush counters before exit (CLI has no FPM shutdown spill hook; write shard for scheduler merge)
+        // Flush counters before exit (CLI has no FPM shutdown spill hook; append journal line for scheduler merge)
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply scripts/unified-webcam-worker.php:332]: Same fix as the other exit path - 4e0b73d renames the helper to metrics_write_spill_journal_and_reset_counters().

File: lib/metrics-spill-aggregator.php [hunk 8]
```diff
@@ -118,25 +118,33 @@ function metrics_run_spill_aggregator_once(): array
...
+                    $hourHadMerges = true;
+                    $stats['spills_merged'] += $mergedUnits;
+                    if ($consumedPath !== null) {
+                        $pendingDeletes[] = $consumedPath;
+                    }
+                } elseif ($consumedPath !== null) {
+                    if (@unlink($consumedPath)) {
+                        $stats['spills_deleted']++;
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-aggregator.php]: Good catch. Returning 0 on fopen/flock failure was ambiguous and still counted against the per-run file budget even though we already guarded deletion with journalFullyConsumed. Fixed in 00b52fc: merge_claimed now returns null on open/lock failure, merge_journal propagates that, and the aggregator continues without deleting or incrementing filesMergedThisRun.

File: lib/metrics-spill-aggregator.php [hunk 9]
```diff
@@ -227,108 +235,65 @@ function metrics_spill_aggregator_write_hour_bucket(
...
     }
 
-    return metrics_parse_spill_payload_for_merge($data, $expectedHourId);
+    return null;
 }
 
 /**
- * Remove stale spill artifacts that were never merged (legacy shards, journals, abandoned claims).
+ * Remove stale spill journals, abandoned claim files, and legacy pre-JSONL .json shards.
  *
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-aggregator.php:332]: Fair point - we removed legacy merge, not legacy orphan cleanup. Updated the PR description to call out that stale pre-JSONL .json shards are still pruned by metrics_spill_aggregator_prune_orphan_spills as a deploy safety net (delete only, no merge).

File: tests/Unit/MetricsSpillAggregatorOrphanPruneTest.php [hunk 10]
```diff
@@ -13,16 +13,16 @@
...
         if (!@mkdir($dir, 0755, true) && !is_dir($dir)) {
             $this->fail('Could not create spill hour directory');
         }
 
-        $stalePath = $dir . '/1.json';
-        $this->assertNotFalse(file_put_contents($stalePath, json_encode(['orphan' => true])));
+        $stalePath = getMetricsSpillWorkerJournalPath($oldHour, 1);
+        $this->assertNotFalse(file_put_contents($stalePath, "{\"orphan\":true}\n"));
         $staleMtime = time() - METRICS_SPILL_ORPHAN_MAX_AGE_SECONDS - 120;
         $this->assertTrue(touch($stalePath, $staleMtime));
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillAggregatorOrphanPruneTest.php]: Agreed - valid hour dirs were getting claimed in the merge loop so spills_deleted could fire instead of orphans_pruned. Fixed in 8ffa4c5: stale artifacts now live under invalid hour directory names the merge loop skips.

File: lib/metrics-spill-journal.php [hunk 11]
```diff
@@ -101,30 +107,49 @@ function metrics_spill_journal_claim_for_merge(string $journalPath): ?string
...
+            ) {
+                $budgetExhausted = true;
+                if ($merged > 0) {
+                    metrics_spill_journal_rewrite_claimed_tail($fp, $claimedPath);
+                }
+                break;
+            }
+
             $line = trim($rawLine);
             if ($line === '') {
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply lib/metrics-spill-journal.php:177]: Agreed - this was a real data-loss edge case. Fixed in 7e1cd8f (cherry-picked from #116 c290f3c): when the budget trips after fgets(), the unread rawLine is prepended to the rewritten claim tail so no delta is dropped.

File: tests/Unit/MetricsSpillJournalTest.php [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+use PHPUnit\Framework\TestCase;
+
+require_once __DIR__ . '/../../lib/cache-paths.php';
+require_once __DIR__ . '/../../lib/constants.php';
+require_once __DIR__ . '/../../lib/metrics-spill-journal.php';
+
+class MetricsSpillJournalTest extends TestCase
+{
+    public function testBuildPayload_IncludesSchemaHourIdAndCounters(): void
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillJournalTest.php]: Fair point - aligned with CODE_STYLE. Fixed in the latest commit: renamed to testMetricsSpillBuildPayload_IncludesSchemaHourIdAndCounters.

File: tests/Unit/MetricsSpillJournalTest.php [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        $payload = metrics_spill_build_payload($hourId, 88020, ['global_page_views' => 9]);
+
+        $this->assertSame(METRICS_SPILL_FILE_SCHEMA_VERSION, $payload['schema_version']);
+        $this->assertSame($hourId, $payload['hour_id']);
+        $this->assertSame(88020, $payload['pid']);
+        $this->assertSame(9, $payload['counters']['global_page_views']);
+        $this->assertIsInt($payload['generated_at']);
+    }
+
+    public function testAppendLocked_AppendsNewlineTerminatedLines(): void
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillJournalTest.php]: Agreed. Renamed to testMetricsSpillJournalAppendLocked_AppendsNewlineTerminatedLines in the same commit.

File: tests/Unit/MetricsSpillJournalTest.php [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+        $raw = file_get_contents($journal);
+        $this->assertNotFalse($raw);
+        $this->assertSame(2, substr_count($raw, "\n"));
+
+        @unlink($journal);
+        @rmdir($dir);
+    }
+
+    public function testClaimForMerge_RenamesLiveJournalAwayFromAppendPath(): void
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillJournalTest.php]: Agreed. Renamed to testMetricsSpillJournalClaimForMerge_RenamesLiveJournalAwayFromAppendPath in the same commit.

File: tests/Unit/MetricsSpillJournalTest.php [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        $this->assertFileDoesNotExist($live);
+        $this->assertFileExists($claimed);
+        $this->assertTrue(metrics_spill_path_is_claimed_journal($claimed));
+        $this->assertFalse(metrics_spill_path_is_worker_journal($claimed));
+
+        @unlink($claimed);
+        @rmdir($dir);
+    }
+
+    public function testPathClassification_DistinguishesLiveClaimedAndTmp(): void
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillJournalTest.php]: Good call - both helpers are under test here. Renamed to testMetricsSpillPathIsWorkerJournal_MetricsSpillPathIsClaimedJournal_DistinguishesLiveClaimedAndTmp in the same commit.

File: tests/Unit/MetricsSpillJournalTest.php [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        $this->assertTrue(metrics_spill_path_is_worker_journal('/cache/spill/2099-01-01-01/42.jsonl'));
+        $this->assertFalse(metrics_spill_path_is_worker_journal('/cache/spill/2099-01-01-01/42.jsonl.merging.9'));
+        $this->assertFalse(metrics_spill_path_is_worker_journal('/cache/spill/2099-01-01-01/42.jsonl.tmp.1'));
+
+        $this->assertTrue(metrics_spill_path_is_claimed_journal('/cache/spill/2099-01-01-01/42.jsonl.merging.9'));
+        $this->assertFalse(metrics_spill_path_is_claimed_journal('/cache/spill/2099-01-01-01/42.jsonl'));
+    }
+
+    public function testListJournalPathsForHour_IncludesAbandonedClaimFiles(): void
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillJournalTest.php]: Agreed. Renamed to testMetricsSpillAggregatorListJournalPathsForHour_IncludesAbandonedClaimFiles in the same commit.

File: tests/Unit/MetricsSpillAggregatorTest.php [hunk 17]
```diff
@@ -9,46 +9,234 @@
...
+        $this->seedJournal($hourId, 88011, ['global_page_views' => 3]);
+
+        $hourlyDir = CACHE_METRICS_HOURLY_DIR;
+        $previousMode = @fileperms($hourlyDir);
+        if ($previousMode === false) {
+            $this->markTestSkipped('Could not read hourly metrics directory permissions');
+        }
+
+        @chmod($hourlyDir, 0555);
+
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillAggregatorTest.php]: Agreed - chmod can fail silently with @ and leave the write succeeding. Fixed in f44a6b7: skip the test when we cannot make the hourly directory read-only.

File: tests/Unit/MetricsSpillAggregatorTest.php [hunk 18]
```diff
@@ -9,46 +9,234 @@
...
+        try {
+            $stats = metrics_run_spill_aggregator_once();
+            $this->assertGreaterThanOrEqual(1, (int) $stats['spills_merged']);
+            $this->assertSame(0, (int) $stats['spills_deleted']);
+            $this->assertFileDoesNotExist(getMetricsHourlyPath($hourId));
+
+            $claimed = glob(getMetricsSpillHourDir($hourId) . '/88011.jsonl.merging.*') ?: [];
+            $this->assertCount(1, $claimed);
+        } finally {
+            @chmod($hourlyDir, $previousMode);
```
[the remark this answers is not in the document]
Comment (alexwitherspoon) [reply tests/Unit/MetricsSpillAggregatorTest.php:236]: Good point. Fixed in c904204: mask fileperms() to 0777 before restoring the hourly directory mode.

**Comment by alexwitherspoon** [comment]
This won't be merged untill Jun 10th.
